Source/WebGPU/WGSL/GlobalVariableRewriter.cpp

```diff
@@ -355,6 +355,125 @@ void RewriteGlobalVariables::visitEntryPoint(AST::Function& function, AST::Stage
     insertLocalDefinitions(function, usedGlobals.privateGlobals);
 }
 
+static BindGroupLayoutEntry::BindingMember bindingMemberForGlobal(auto& global)
+{
+    auto* variable = global.declaration;
+    ASSERT(variable);
+    auto* type = variable->storeType();
+    ASSERT(type);
+
+    using namespace WGSL::Types;
+    return WTF::switchOn(*type, [&](const Primitive& primitive) -> BindGroupLayoutEntry::BindingMember {
+        switch (primitive.kind) {
+        case Types::Primitive::AbstractInt:
+        case Types::Primitive::I32:
+        case Types::Primitive::U32:
+        case Types::Primitive::AbstractFloat:
+        case Types::Primitive::F32:
+        case Types::Primitive::Void:
+        case Types::Primitive::Bool:
+            RELEASE_ASSERT_NOT_REACHED();
+        case Types::Primitive::Sampler:
+            return SamplerBindingLayout {
+                .type = SamplerBindingType::Filtering
+            };
+        case Types::Primitive::TextureExternal:
+            return ExternalTextureBindingLayout { };
+        }
+    }, [&](const Vector& vector) -> BindGroupLayoutEntry::BindingMember {
+        auto* primitive = std::get_if<Primitive>(vector.element);
+        UNUSED_PARAM(primitive);
+        return BufferBindingLayout {
+            .type = BufferBindingType::Uniform,
+            .hasDynamicOffset = false,
+            .minBindingSize = 0
+        };
+    }, [&](const Matrix& matrix) -> BindGroupLayoutEntry::BindingMember {
+        UNUSED_PARAM(matrix);
+        return BufferBindingLayout {
+            .type = BufferBindingType::Uniform,
+            .hasDynamicOffset = false,
+            .minBindingSize = 0
+        };
+    }, [&](const Array& array) -> BindGroupLayoutEntry::BindingMember {
+        UNUSED_PARAM(array);
+        return BufferBindingLayout {
+            .type = BufferBindingType::Uniform,
+            .hasDynamicOffset = false,
+            .minBindingSize = 0
+        };
+    }, [&](const Struct& structure) -> BindGroupLayoutEntry::BindingMember {
+        UNUSED_PARAM(structure);
+        return BufferBindingLayout {
+            .type = BufferBindingType::Uniform,
+            .hasDynamicOffset = false,
+            .minBindingSize = 0
+        };
```

Comment:
Unconditionally saying `.type = BufferBindingType::Uniform` isn't right. The address space is supposed to determine that:

```
@group(0) @binding(0) var<uniform> lights : LightStorage;
```

vs

```
@group(0) @binding(0) var<storage> lights : LightStorage;
```

Reply:
Oh good point, let me see how to get the addressSpace out of a `WGSL::Structure`. @tadeuzagallo or @djg do you happen to know off hang?